### ISSUE TITLE
Recurrent Drafter not working

### ISSUE THREAD

**Issue by binhtranmcs** [NVIDIA/TensorRT-LLM#2155]
### System Info

- GPU: nvidia A30
- TensorRT-LLM: commit [32ed92e](https://github.com/chiendb97/TensorRT-LLM/commit/32ed92e4491baf2d54682a21d247e1948cca996e)
- Nvidia driver: 535.86.10
- Ubuntu 22.04.4

### Who can help?

@byshiue @kaiyux 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Steps to reproduce:
- Follow the steps in this [readme](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/redrafter) to build tensorrt engine.
- Run test using `run.py` script with flags `--streaming --streaming_interval 1`

### Expected behavior

Multiple tokens are generated at the same time at each step

### actual behavior

Tokens are generated one by one, and it is much slower than normal model

### additional notes

I tested with the script provided [here](https://github.com/apple/ml-recurrent-drafter?tab=readme-ov-file#run-the-recurrent-drafter) and multiple tokens are generated at the same time as expected. Token/step is around 3.9, so I think it is a bug when convert/build/run the tensorrt llm engine. Please have a look.

**Comment by rakib-hasan** [comment]
Hi @binhtranmcs . I can't seem to reproduce the issue on my end.
Could you please share the exact commands you used for convert/build/run? Also, which model and drafter did you use?

**Comment by binhtranmcs** [comment]
@rakib-hasan, I tested with vicuna-7b and it works fine. But when using my own llama3-8b, it did not work as expected. I cannot provide you the model, so I will try to debug and update with you later.

**Comment by rakib-hasan** [comment]
@binhtranmcs Thanks for the information. I understand. I was actually asking for the details of your experiment like how you did training and then convert, build, and run.
Please let us know once you have more information that you can share.

**Comment by binhtranmcs** [comment]
@rakib-hasan, I use the script [here](https://github.com/apple/ml-recurrent-drafter/blob/main/recurrent_drafting/cmd/train.sh) to train on 4 H100 gpus. The benchmark result: ![Image](https://github.com/user-attachments/assets/57655847-6006-42cc-854d-2d3bd3f344d3)



Then I follow the steps [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/redrafter) to convert and run model.

Convert model:
```
python examples/redrafter/convert_checkpoint.py \
  --model_dir MODEL_DIR \
  --drafter_model_dir DRAFTER_DIR \
  --output_dir CONVERT_DIR \
  --dtype float16 \
  --redrafter_num_beams 45 \
  --redrafter_draft_len_per_beam 5 \
  --tp_size 1
```
Build engine:
```
trtllm-build \
  --checkpoint_dir CONVERT_DIR \
  --output_dir ENGINE_DIR \
  --gemm_plugin float16 \
  --gpt_attention_plugin float16 \
  --speculative_decoding_mode explicit_draft_tokens \
  --max_batch_size 2 \
  --use_paged_context_fmha enable \
  --context_fmha enable --paged_kv_cache enable \
  --remove_input_padding enable \
  --max_beam_width 1 --max_input_len 1024 --max_seq_len 1024
```
Run test:
```
mpirun -np 1 --allow-run-as-root python3 examples/run.py \
  --engine_dir ENGINE_DIR \
  --tokenizer_dir TOKENIZER_DIR \
  --max_output_len 50 \
  --input_text TEXT \
  --streaming --streaming_interval=1
```

**Comment by rakib-hasan** [comment]
@binhtranmcs One possible reason for the bad performance is probably the `redrafter_num_beams` being so high. 

With num_beams `45` and draft length `5`, for each sequence, in each step, maximum possible number of tokens during generation phase is `45*(5+1) = 270`. Because of prefix sharing, if it is 50% of that on average, you are still processing `~135` tokens in each step compared to `1` token when it is base model only.
Now if we assume: 

1. processing `1` token vs `32` tokens takes about the same amount of time
2. after `32` tokens, the execution time linearly increases

Then with processing `135` tokens in each step, step latency would increase by `~4x`. Based on the data you shared, your average acceptance rate is `~3`. So, it would lead to slow down even for `BS=1`.

**Comment by binhtranmcs** [comment]
@rakib-hasan, I understand that, but I am not worrying about performance for now. The problem I am facing is that when running llama3-8b model with redrafter, only one output token is generated per step, whereas the expected behaviour is to have multiple ones generated. Also, prior to conversion to tensorrt_llm, running model using [this](https://github.com/apple/ml-recurrent-drafter?tab=readme-ov-file#run-the-recurrent-drafter) generates multiple tokens per step. So I think it might be a bug of tensorrt_llm when running redrafter internally.

**Comment by rakib-hasan** [comment]
I see. Unfortunately, I think run.py doesn't have a way to show that data yet.
If you modify model_runner_cpp.py to print the iteration stats from executor, it should confirm whether more than 1 token is getting accepted or not in each iteration. If it does show that it is only accepting 1 token at each step, you might want to try the debug mode which will save the IO tensors at every step which should help debug the issue. Hopefully this helps.

**Comment by binhtranmcs** [comment]
You can use `run.py` with args `--streaming --streaming_interval=1` as above to print out the output length of every step:
```
    if args.streaming:
        for curr_outputs in throttle_generator(outputs,
                                               args.streaming_interval):
            if runtime_rank == 0:
                output_ids = curr_outputs['output_ids']
                sequence_lengths = curr_outputs['sequence_lengths']
                print("debug", sequence_lengths)
```
 With llama3-8b, the output length only increases by 1 each step. But with vicuna-7b, the output length may increase by 2 or 3. That is how I figured out that the redrafter for llama3-8b was not working. Maybe you can try to reproduce with the llama3-8b model.

**Comment by rakib-hasan** [comment]
You are right. That would an alternative to see that data.
As for trying llama3-8b, I don't have a trained drafter for it yet. Not sure if I will be able to train one any time soon.

**Comment by rakib-hasan** [comment]
@binhtranmcs were you able to figure out the issue with llama3-8b redrafter? Please let us know if we can help fix any issues on our side.

**Comment by binhtranmcs** [comment]
@rakib-hasan it was some issues with our training. The redrafter seems to be working now.
But when running benchmarks, the performance is worse than running the base model alone. Do you have a benchmark results when running **_any open source model_** with redrafter? Some metrics like latency and throughput when running server with inflight batching would be enough. In the meantime, I will try to tune the parameters to achieve better results.

**Comment by rakib-hasan** [comment]
@binhtranmcs Unfortunately, I am not aware of any open-source model with ReDrafter at the moment. If you know of any, please let us know.

However, we did find 2 bugs very recently that can affect performance and could be the case for you. 
1. If the number of beams is too high (>10), topk is running very slowly (I think you were running into this with num_beams=45). I am looking into possible workarounds for that until we have a proper fix. 
2. There is a bug in TensorRT that could cause beam search accuracy to be very low, thus leading to poor acceptance rate. This behavior can vary from model to model. This bug should be fixed in TRT-10.6 .

**Comment by binhtranmcs** [comment]
I just tested again with llama3-8B model, trained on alpaca dataset with `redrafter_num_beams=6`,  following these steps:
- Convert checkpoint:
```
python examples/redrafter/convert_checkpoint.py \
  --model_dir $MODEL_DIR \
  --drafter_model_dir $DRAFTER_DIR \
  --output_dir $CONVERT_DIR \
  --dtype float16 \
  --redrafter_num_beams 6 \
  --redrafter_draft_len_per_beam 5 \
  --tp_size 2
```
- Build engine:
```
trtllm-build \
  --checkpoint_dir $CONVERT_DIR \
  --output_dir $ENGINE_DIR \
  --gemm_plugin float16 \
  --gpt_attention_plugin float16 \
  --speculative_decoding_mode explicit_draft_tokens \
  --max_batch_size 64 \
  --use_paged_context_fmha enable \
  --context_fmha enable --paged_kv_cache enable \
  --remove_input_padding enable \
  --max_beam_width 1 --max_input_len 2048 --max_seq_len 2048
```
- Deploy with tritonserver using tensorrtllm_backend, with tp_size=2 and pp_size=1.
- Benchmark the server using this script [benchmark.txt](https://github.com/user-attachments/files/17410551/benchmark.txt):
```
python benchmark.py  -i grpc --decoupled \
  --max-input-len 2048 --exclude-input-in-output --request-rate 10 --num-requests 1000 \
  dataset --dataset $DATASET_PATH \
  --tokenizer-dir $MODEL_DIR --tokenizer-type llama
```

I got the benchmark result:
```
+----------------------------+-----------+
|            Stat            |   Value   |
+----------------------------+-----------+
|        Requests/Sec        |   2.57    |
|       OP tokens/sec        |  131.25   |
|     Avg. latency (ms)      | 136822.32 |
|      P99 latency (ms)      | 272968.58 |
|      P90 latency (ms)      | 242824.32 |
| Avg. IP tokens per request |   47.66   |
| Avg. OP tokens per request |   51.07   |
|   Avg. InFlight requests   |   0.00    |
|     Total latency (ms)     | 389095.54 |
|       Total requests       |  1000.00  |
+----------------------------+-----------+
```

Running the base model alone without redrafter:
```
+----------------------------+-----------+
|            Stat            |   Value   |
+----------------------------+-----------+
|        Requests/Sec        |   9.31    |
|       OP tokens/sec        |  475.33   |
|     Avg. latency (ms)      |  693.28   |
|      P99 latency (ms)      |  2962.41  |
|      P90 latency (ms)      |  1548.28  |
| Avg. IP tokens per request |   47.66   |
| Avg. OP tokens per request |   51.07   |
|   Avg. InFlight requests   |   0.00    |
|     Total latency (ms)     | 107438.21 |
|       Total requests       |  1000.00  |
+----------------------------+-----------+
```

You can see that the throughput of the redrafter is much worse than the base model.

@rakib-hasan you can give me your huggingface email and username, I will share with you the base model with the redrafter so you can reproduce and run benchmark on your own.

**Comment by rakib-hasan** [comment]
@binhtranmcs Thanks for the details. I am not sure if I am allowed to post my email address publicly on GitHub. The username is **rh-nv** in case that works. If not, if you can share the HF link to the model, I can try to request access.

**Comment by binhtranmcs** [comment]
@rakib-hasan my colleague just sent the invitation to "ZLB AI", please check your email.

**Comment by rakib-hasan** [comment]
Thanks @binhtranmcs . I was able to access the model and reproduce the bad acceptance rate issue. Looking into this.

**Comment by rakib-hasan** [comment]
@binhtranmcs 
It seems the bad acceptance rate is just due to the model and the task I was running.

I built the TRT-LLM engine with 2 different configs and ran with a single prompt using the `examples/run.py` script.
The input prompt: `1000 words essay on Tourism would go like this : `
To compare the output and acceptance rate, I also ran the model with PyTorch using [redrafter repo](https://github.com/apple/ml-recurrent-drafter) `cmd/generate.py`. In both PyTorch and TRT-LLM cases, I recorded the number of accepted tokens at each step.

 **Config: `BW=4 BL=5 (i.e. num_beams=4 and draft_len_per_beam=4)`**

PyTorch:
```
Accepted tokens: 1,5,5,1,1,1,1,3,3,1,1,1,5,1,1,4,3,1,1,3,2,3,2,1,1,1,1,1,1,3,1,1,1,1,1,4,4,2,2,1,1,1,1,2,2,1,1,1,2,1,2,3,1,1,3,1,1,4,3,1,1,1,3,2,1,4,1,3,1,2,1,1,1,2,1,1,1,2,2,4,2,1,3,1,4,1,1,5,1,2,1,5,2,1,4,
" Tourism is one of the most important industries in the world, bringing in billions of dollars each year and providing jobs for millions of people. It has a huge impact on the global economy, and has the potential to be a great source of revenue for countries and cities that have attractive tourist destinations.

Tourism can also be a great way for people to learn more about other cultures, and to build connections and understanding between different countries and communities. Tourists can gain valuable insights and experiences when they travel, and this can help to create a better world by fostering respect and understanding between different peoples.

In conclusion, tourism can be a great way to generate revenue for destinations and provide jobs for locals, while also providing an opportunity to learn more about different cultures and build connections between people. It has the potential to be a powerful force for good in the world, and should be encouraged wherever possible.<|end_of_text|>"
Acceptance Rate: 178 / 95 = 1.87 tokens/step
```
TRT-LLM:
```
Accepted tokens: 1,5,5,1,1,1,1,3,3,1,1,1,5,1,1,4,3,1,1,3,2,3,2,1,1,1,1,1,1,3,1,1,1,1,1,4,4,2,2,1,1,1,1,2,2,1,1,1,2,1,2,3,1,1,3,1,1,4,3,1,1,1,3,2,1,4,1,3,1,2,1,1,1,2,1,1,1,2,2,4,2,1,3,1,4,1,1,5,1,2,1,5,2,1,4,
" Tourism is one of the most important industries in the world, bringing in billions of dollars each year and providing jobs for millions of people. It has a huge impact on the global economy, and has the potential to be a great source of revenue for countries and cities that have attractive tourist destinations.

Tourism can also be a great way for people to learn more about other cultures, and to build connections and understanding between different countries and communities. Tourists can gain valuable insights and experiences when they travel, and this can help to create a better world by fostering respect and understanding between different peoples.

In conclusion, tourism can be a great way to generate revenue for destinations and provide jobs for locals, while also providing an opportunity to learn more about different cultures and build connections between people. It has the potential to be a powerful force for good in the world, and should be encouraged wherever possible.<|end_of_text|>"
Acceptance Rate: 178 / 95 = 1.87 tokens/step
```
You can see that the output and accepted tokens at each step is identical for this config. 


**Config: `BW=6 BL=6 (i.e. num_beams=6 and draft_len_per_beam=5)`**

PyTorch:
```
Accepted tokens: 1,6,1,3,1,1,1,1,3,3,1,6,3,4,3,1,1,3,2,3,2,1,1,1,1,1,1,3,1,1,1,1,1,4,4,2,2,1,1,1,1,2,2,1,1,1,2,1,2,3,1,1,3,1,1,4,3,1,1,1,6,2,2,2,3,4,2,1,2,1,4,2,2,2,3,5,4,1,1,2,2,1,2,1,2,4,1,3,4,2,1,4,1,1,1,4,2,1,5,2,6,3,
" Tourism is one of the most important industries in the world, bringing in billions of dollars each year and providing jobs for millions of people. It has a huge impact on the global economy, and has the potential to be a great source of revenue for countries and cities that have attractive tourist destinations.

Tourism can also be a great way for people to learn more about other cultures, and to build connections and understanding between different countries and communities. Tourists can gain valuable insights and experiences when they travel, and this can help to create a better world by fostering respect and understanding between different cultures.

Furthermore, tourism can have a positive economic impact on local communities. By investing in tourist infrastructure, such as hotels and attractions, money is injected into the local economy, which can create new jobs and increase the standard of living for the locals.

Overall, tourism is a great industry, with the potential to create economic benefits and understanding between different cultures. It is important that we continue to invest in tourism, so that it can continue to have a positive impact on the world.<|end_of_text|>"
Acceptance Rate: 215 / 102 = 2.11 tokens/step
```

TRT-LLM:
```
Accepted tokens: 1,6,1,3,1,1,1,1,3,3,1,6,3,4,3,1,1,3,2,3,2,1,1,1,1,1,1,3,1,1,1,1,1,4,4,2,2,1,1,1,1,2,2,1,1,1,2,1,2,3,1,1,3,1,1,4,3,1,1,1,5,1,4,1,3,1,2,1,1,1,2,1,1,1,2,2,4,2,4,1,4,1,6,1,1,1,4,2,2,1,4,
" Tourism is one of the most important industries in the world, bringing in billions of dollars each year and providing jobs for millions of people. It has a huge impact on the global economy, and has the potential to be a great source of revenue for countries and cities that have attractive tourist destinations.

Tourism can also be a great way for people to learn more about other cultures, and to build connections and understanding between different countries and communities. Tourists can gain valuable insights and experiences when they travel, and this can help to create a better world by fostering respect and understanding between different peoples.

In conclusion, tourism can be a great way to generate revenue for destinations and provide jobs for locals, while also providing an opportunity to learn more about different cultures and build connections between people. It has the potential to be a powerful force for good in the world, and should be encouraged wherever possible.<|end_of_text|>"
Acceptance Rate: 178 / 91 = 1.96 tokens/step
```

For this config, TRT-LLM output is still identical to the first config as expected (though the acceptance rate is slightly better due to wider beam). But PyTorch started diverging a bit (seems to be colliding probabilities) and continues to generate more tokens.

The above experiment was done without the bug that I mentioned earlier in the beam search. However, I also ran the first config with the bug present, which doesn't change the acceptance rate much (1.87 vs 1.89). 
Based on this, here are my observations so far:
1. This specific model that you shared doesn't seem to be affected by the beam search bug.
2. TRT-LLM implementation seems to be correctly producing the same beams and outputs as PyTorch.

As for the bad performance that you shared in your last experiment, upon looking at the build and run script, I think I understand why.
In the latest data you shared, your `max_batch_size` is `64` which is too high for any speculative decoding technique. At that batch size, the model is most likely running at the compute-bound region. In that case, doing speculative decoding is likely hurt performance due to wasted computation. 
Speculative decoding techniques should only be applied when the GPU is underutilized (memory-bound) due to batch size being low i.e. very low traffic.

I hope this makes sense. Please let me know if you have any more questions.

**Comment by rakib-hasan** [comment]
@binhtranmcs Today's update includes fixes for both the beam search accuracy bug and the topK performance. 

You can try it and see if that helps with your usecase.
Please keep in mind that while topK is now improved for larger number of ReDrafter beams, if you use a value that is too large, the performance might still suffer due to the increased number of draft tokens per step. Also, please remember to use `max_batch_size=1` for the best performance improvements.

**Comment by binhtranmcs** [comment]
@rakib-hasan thanks for your insights and support. The redrafter seems to run fine now. Will close this issue.
